ptclient/http.c

```diff
@@ -0,0 +1,396 @@
+    1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, /* 30-3F */
+
+    1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, /* 40-4F */
+    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, /* 50-5F */
+    1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, /* 60-6F */
+    2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 0, /* 70-7F */
+
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+};
+
+/*
+ * Convert 'identifier' into canonical form.
+ * Returns a pointer to a static buffer containing the canonical form
+ * or NULL if 'identifier' is invalid.
+ *
+ * XXX If any of the characters marked with 0 are valid and are cropping up,
+ * the right thing to do is probably to canonicalize the identifier to two
+ * representations: one for getpwent calls and one for folder names.  The
+ * latter canonicalizes to a MUTF7 representation.
+ */
+static char *mycanonifyid(const char *identifier, size_t len)
+{
+    static char retbuf[81];
+    char sawalpha;
+    char *p;
+    int username_tolower = 0;
+    int i = 0;
+
+    if(!len) len = strlen(identifier);
+    if(len >= sizeof(retbuf)) return NULL;
+
+    memcpy(retbuf, identifier, len);
+    retbuf[len] = '\0';
+
+    if (!strncmp(retbuf, "group:", 6))
+        i = 6;
+
+    /* Copy the string and look up values in the allowedchars array above.
+     * If we see any we don't like, reject the string.
+     * Lowercase usernames if requested.
+     */
+    username_tolower = config_getswitch(IMAPOPT_USERNAME_TOLOWER);
+    sawalpha = 0;
+    for(p = retbuf+i; *p; p++) {
+        if (username_tolower && Uisupper(*p))
+            *p = tolower((unsigned char)*p);
+
+        switch (allowedchars[*(unsigned char*) p]) {
+        case 0:
+            return NULL;
+
+        case 2:
+            sawalpha = 1;
+            /* FALL THROUGH */
+
+        default:
+            ;
+        }
+    }
+
+    if (!sawalpha) return NULL;  /* has to be one alpha char */
+
+    return retbuf;
+}
+
+
+/* API */
+
+static void myinit(void)
+{
+    char *p;
+    size_t n;
+
+    if (state.uri.raw) return; // Already configured
+
+    /* Fetch and parse URI uri */
+    p = state.uri.raw = xstrdupnull(config_getstring(IMAPOPT_HTTPPTS_URI));
+    if (!p) fatal("Missing 'httppts_uri' option", EX_CONFIG);
+
+    if (!strncasecmp(p, "https://", 8)) {
+        state.uri.host = p += 8;
+        state.uri.port = 443;
+        state.uri.https = 1;
+    }
+    else if (!strncasecmp(p, "http://", 7)) {
+        state.uri.host = p += 7;
+        state.uri.port = 80;
+    }
+    else fatal("Invalid 'httppts_uri' scheme", EX_CONFIG);
```

Comment:
is the some URI parser library we could/should be using here?

Reply:
I tried using xmlParseURI, which we use in httpd.c, but it doesn't like the {} in the template.  I didn't want to add another dependency for something that is easily parsable